### PR TITLE
Create generic DB dynamic proxy interceptor, intercept CockroachDB

### LINKED ISSUE #156: Exceptions for Cockroach DB
Before moving onto the BFT algorithm -- since we're quite a bit ahead of schedule -- can you see if anything you did for Redis could be used to do the same type of exception-injection for CockroachDB as well?  Then, we might be able to generalize the FI for databases and from there explore BFT strategies in general.  Can you investigate and see what it might take to do what you've done for Redis exception injection for Cockroach?

### PR DESCRIPTION
**Pull request by MichaelA-ops**
Closes #156 

This PR:
- [X] Creates a generic dynamic proxy interceptor that can potentially be used with any database client (Maybe use this interceptor later for Lettuce as well?)
- [X] Wires up the interceptor class with the Filibuster Client Intrumentor
- [X] Adds a CockroachDB service using a Docker TestContainer
- [X] Adds a basic CockroachDB DAO for testing
- [X] Creates a unit test for exception injection on PGSimpleDataSource/CockroachDB
- [X] Introduces a single fault analysis config file for PGSimpleDataSource throwing `PSQLException`
- [X] Refactors the analysis config files

### CODE REVIEW

**Comment by MichaelA-ops** [comment]
@cmeiklejohn There is still some stuff to do, but this PR is already quite big and introduces basic exception injection for CockroachDB as described in #156. I will open new issues/PRs for further exceptions and BFI

File: src/main/java/cloud/filibuster/instrumentation/Constants.java [hunk 1]
```diff
@@ -5,7 +5,9 @@ public class Constants {
     final public static String FILIBUSTER_ORIGIN_VCLOCK = "_filibuster_origin_vclock";
     final public static String FILIBUSTER_REQUEST_ID = "_filibuster_request_id";
     final public static String FILIBUSTER_EXECUTION_INDEX = "_filibuster_execution_index";
-    // Used in the RedisInterceptor and the Redis analysis configuration files. The name "RedisClient" is used as
+
+    // Module names are used in the DB interceptors and analysis configuration files. They are used as
     // classOrModuleName for CallSite objects or for pattern recognition in the analysis configuration files.
     final public static String REDIS_MODULE_NAME = "RedisClient";
+    final public static String COCKROACH_MODULE_NAME = "CockroachClient";
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/Constants.java:12]: Probably pg simple here?

File: src/main/java/cloud/filibuster/instrumentation/Constants.java [hunk 2]
```diff
@@ -5,7 +5,9 @@ public class Constants {
     final public static String FILIBUSTER_ORIGIN_VCLOCK = "_filibuster_origin_vclock";
     final public static String FILIBUSTER_REQUEST_ID = "_filibuster_request_id";
     final public static String FILIBUSTER_EXECUTION_INDEX = "_filibuster_execution_index";
-    // Used in the RedisInterceptor and the Redis analysis configuration files. The name "RedisClient" is used as
+
+    // Module names are used in the DB interceptors and analysis configuration files. They are used as
     // classOrModuleName for CallSite objects or for pattern recognition in the analysis configuration files.
     final public static String REDIS_MODULE_NAME = "RedisClient";
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/Constants.java:11]: We could also make this a bit more precise with lettuce too.

**Comment by cmeiklejohn** [comment]
Please change CHANGELOG and build.gradle version.

File: src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static cloud.filibuster.instrumentation.helpers.Property.*;
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java:26]: Again, no wildcard imports.  Fix your editor yet? :)

File: src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    private static final String logPrefix = "[FILIBUSTER-PROXY_INTERCEPTOR]: ";
+    private final String moduleName;
+    private FilibusterClientInstrumentor filibusterClientInstrumentor;
+
+    private DynamicProxyInterceptor(T targetObject, String connectionString, String moduleName) {
+        logger.log(Level.INFO, logPrefix + "Constructor was called");
+        this.targetObject = targetObject;
+        this.contextStorage = new ThreadLocalContextStorage();
+        this.connectionString = connectionString;
+        this.serviceName = getServiceName(connectionString);
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java:49]: Which one is this calling?  The static or the non-static?  It's not clear when reading.

File: src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        this.moduleName = moduleName;
+    }
+
+    public String getServiceName() {
+        return this.serviceName;
+    }
+
+    private static String getServiceName(String connectionString) {
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java:57]: Might want to give this a different name, I didn't realize it was overloaded static vs non-static with the same name.

File: src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        // ******************************************************************************************
+        // Construct preliminary call site information.
+        // ******************************************************************************************
+
+        Class<?> argsClass = args != null ? args.getClass() : Object[].class;
+        String argsString = args != null ? Arrays.toString(args) : "[]";
+
+        CallsiteArguments callsiteArguments = new CallsiteArguments(argsClass, argsString);
+
+        Callsite callsite = new Callsite(serviceName, moduleName, fullMethodName, callsiteArguments);
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java:93]: Can you give examples here of what the method name vs. call site arguments (in comments) will be so we can make sure it's correct?

File: src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java (hunk 3 above)
Comment (MichaelA-ops) [reply src/main/java/cloud/filibuster/instrumentation/libraries/dynamic/proxy/DynamicProxyInterceptor.java]: Seems like I only fixed my Surface IDE. Now fixed it on my Macbook as well

**Comment by MichaelA-ops** [comment]
@cmeiklejohn The generic proxy interceptor seems to be working for DynamoDB as well! I will make some minor adjustment and probably submit a PR today. https://github.com/filibuster-testing/filibuster-java-instrumentation/tree/158-create-dynamodb-interceptor

**Comment by cmeiklejohn** [comment]
I didn't get a chance to review this before you merged, I only changed the base branch.

**Pull request merged**